MoreLinq/Subsets.cs

```diff
@@ -108,7 +108,9 @@ public static IEnumerable<IList<T>> Subsets<T>(this IEnumerable<T> sequence, int
             // preconditions. This however, needs to be carefully considered - and perhaps
             // may change after further thought and review.
 
-            return new SubsetGenerator<T>(sequence, subsetSize);
+            return subsetSize == 0
+                ? (IEnumerable<IList<T>>) new[] {new List<T>()}
```

Comment:
Return a cached empty array as opposed to create an empty list each time. While [.NET offers ones](https://docs.microsoft.com/en-us/dotnet/api/system.array.empty?view=netcore-3.1#System_Array_Empty__1), I think we should introduce our own to avoid adding an additional targets (as it's available since .NET 4.6 and .NET Standard 1.3):

```c#
static class EmptyArray<T>
{
    public static readonly T[] Value = new T[0];
}
```

Reply:
It's not an empty array. It's an array of one element. This element is an empty list.

I don't believe we can have a cache for that.

If we use a cache, obviously the list must be immutable / read-only.

In other cases, the method return an `IEnumerable<IList<T>>` where the elements are not read-only list.

Reply:
I made SubsetGenerator handle the k=0 case

Reply:
I meant you can cache the _empty set_, which is a zero element array that's immutable anyway; then you could have returned another cached sequence of that. Anyway, you have changed `SubsetGenerator` to handle _k_ = 0 case so this point is mute. I think those optimizations would have been rarely used and not any better than the current implementation that returns an empty list followed by subsets as arrays for all _k_ where _k_ > 0 (which is inconsistent). I think we can add the empty set optimization in another PR.

Reply:
https://github.com/morelinq/MoreLINQ/blob/07bd0861658b381ce97c8b44d3b9f2cd3c9bf769/MoreLinq/Subsets.cs#L165

`SubsetEnumerator.Current` clone the current state (an array) and return it as an `IList`. 😨

Arrays as `IList` are not read-only, _generally_ not immutable (you can change the value of an element), but are not fully mutable, you can't add or remove elements.

Indeed in the case of the empty array, we have a _de facto_ immutable `IList`, but `IsReadonly` return false.

I think this should be rethought.

My opinion is that it should return a `List`.

If we look at other methods that returns `IEnumerable<IList<T>>` or `IEnumerable<IEnumerable<T>>`,  `Subsets` is the only method that clone the current state when `Enumerator.Current` is called.

```C#
void Study(IEnumerable<IEnumerable<int>> lists)
{
    var e = lists.GetEnumerator();
    e.MoveNext();
    var current = e.Current;
    if (current is List<int>)
        Console.Write("List  ");
    else if (current is int[])
        Console.Write("Array ");
    else
        Console.Write("????  ");

    var otherCurrent = e.Current;

    Console.WriteLine(ReferenceEquals(current, otherCurrent) ? "same ref" : "new ref");
}

var r = Enumerable.Range(0, 10);
var rr = Enumerable.Repeat(r, 10);
Study(r.Batch(3));                 // Array same ref
Study(r.Permutations());           // Array same ref
Study(r.Segment(v => v % 2 == 1)); // List  same ref
Study(r.Split(v => v % 2 == 1));   // List  same ref
Study(r.Subsets(5));               // Array new ref
Study(rr.Transpose());             // Array same ref
Study(r.Window(5));                // Array same ref
```

Reply:
Yes, some inconsistencies have slipped through the cracks over the years.